cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operations/OperationRunsStore.java

```diff
@@ -0,0 +1,384 @@
+            new SimpleImmutableEntry<String, OperationRunDetail<?>>(
+                row.getString(StoreDefinition.OperationRunsStore.ID_FIELD),
+                GSON.fromJson(row.getString(StoreDefinition.OperationRunsStore.DETAILS_FIELD),
+                    OperationRunDetail.class)
+            );
+        keepScanning = func.apply(scanEntry);
+        limit--;
+      }
+    }
+  }
+
+  private Collection<Field<?>> getFilterIndexes(OperationRunFilter filter) {
+    Collection<Field<?>> filterIndexes = new ArrayList<>();
+    if (filter == null) {
+      return filterIndexes;
+    }
+    if (filter.getOperationType() != null) {
+      filterIndexes.add(Fields.stringField(StoreDefinition.OperationRunsStore.TYPE_FIELD,
+          filter.getOperationType()));
+    }
+    if (filter.getStatus() != null) {
+      filterIndexes.add(Fields.stringField(StoreDefinition.OperationRunsStore.STATUS_FIELD,
+          filter.getStatus().toString()));
+    }
+    return filterIndexes;
+  }
+
+  private List<Field<?>> getRangeFields(String namespace, String id)
+      throws IOException, OperationRunNotFoundException {
+    List<Field<?>> fields = new ArrayList<>();
+    fields.add(
+        Fields.stringField(StoreDefinition.OperationRunsStore.NAMESPACE_FIELD, namespace));
+    fields.add(Fields.stringField(StoreDefinition.OperationRunsStore.ID_FIELD, id));
+    Long startTime = getStartTime(new OperationRunId(namespace, id));
+    fields.add(
+        Fields.longField(StoreDefinition.OperationRunsStore.START_TIME_FIELD, startTime));
+    return fields;
+  }
+
+  private OperationRunDetail<?> getCurrentRunDetail(OperationRunId runId)
+      throws IOException, OperationRunNotFoundException {
+    Optional<StructuredRow> currentRow = getOperationRunInternal(runId);
+    if (!currentRow.isPresent()) {
+      throw new OperationRunNotFoundException(runId.getNamespace(), runId.getRun());
+    }
+    return GSON.fromJson(
+        currentRow.get().getString(StoreDefinition.OperationRunsStore.DETAILS_FIELD),
+        OperationRunDetail.class
+    );
+  }
+
+  private long getStartTime(OperationRunId runId)
+      throws IOException, OperationRunNotFoundException {
+    Optional<StructuredRow> currentRow = getOperationRunInternal(runId);
+    if (!currentRow.isPresent()) {
+      throw new OperationRunNotFoundException(runId.getNamespace(), runId.getRun());
+    }
+    return currentRow.get().getLong(StoreDefinition.OperationRunsStore.START_TIME_FIELD);
+  }
+
+  private Collection<Field<?>> getCommonUpdateFields(OperationRunId runId) {
+    Collection<Field<?>> fields = new ArrayList<>();
+    fields.add(
+        Fields.stringField(StoreDefinition.OperationRunsStore.NAMESPACE_FIELD, runId.getNamespace()));
+    fields.add(
+        Fields.stringField(StoreDefinition.OperationRunsStore.ID_FIELD, runId.getRun()));
+    fields.add(
+        Fields.longField(StoreDefinition.OperationRunsStore.UPDATE_TIME_FIELD,
+            System.currentTimeMillis()));
```

Comment:
instead of System.currentTimeMillis(), it would be good to take a Clock in the constructor of this class and use that instead. It would make unit testing simpler, as you can expect a concrete time.

Reply:
Done